collections/_article/whats-new-in-csharp-for-godot-4-0.md

```diff
@@ -0,0 +1,357 @@
+
+The new equivalence tables go into as much detail about what's available, and will hopefully be a great tool for developers looking to port GDScript code to C#. Let us know if this can be improved!
+
+#### Relevant links
+
+- C#: Add table with equivalent string methods ([GH-6442](https://github.com/godotengine/godot-docs/pull/6442)).
+- C#: Add table with equivalent GlobalScope methods ([GH-6721](https://github.com/godotengine/godot-docs/pull/6721)).
+- C#: Add table with equivalent Array methods ([GH-6677](https://github.com/godotengine/godot-docs/pull/6677)).
+- C#: Add table with equivalent Dictionary methods ([GH-6676](https://github.com/godotengine/godot-docs/pull/6676)).
+- Add Color section to C# differences page ([GH-6679](https://github.com/godotengine/godot-docs/pull/6679)).
+
+## NuGet packages
+
+With the Godot 4 release we have also started publishing the Godot assemblies to [NuGet.org](https://www.nuget.org/packages/GodotSharp). This allows developers to create third-party libraries that reference Godot assemblies much more easily.
+
+The NuGet packages versioning follows Godot which uses [semver](https://semver.org/). This means, in theory, that there should not be breaking changes between minor or patch versions. In practice, while we try to avoid breaking compatibility as much as possible, there may still be some minor breaking changes from time to time.
+
+How does breaking compatibility affect your game? If your game references a package _MyLibrary_ which was built against _GodotSharp 4.0.0.0_, but your game is using _GodotSharp 4.1.0.0_ which introduces a breaking change, it will cause runtime exceptions, the most common one being `System.MethodNotFoundException`.
+
+For example, take a look at this method signature:
+
+```csharp
+public void MyMethod();
+```
+
+If this API changes in a future version and the new method signature looks like this:
+
+```csharp
+public void MyMethod(int number = 42);
+```
+
+Even though the parameter is optional, the method is still a different signature therefore libraries built against the previous version that invoke `MyMethod` won't be able to find it.
+
+We will keep trying to get better at avoiding breaking compatibility, but keep in mind there are a number of APIs that we are planning to change in 4.1 that will result in breaking compatibility changes. In order to avoid issues make sure your game project as well as the libraries it uses are all referencing the same _GodotSharp_ version.
```

Comment:
```suggestion
Godot implements various systems to preserve compatibility and allow for smooth transition, even when we have to change some fundamental things. You should expect future releases of Godot 4 to work with your Godot 4.0 project, but we still leave the room for the inevitable. In case of completely incompatible changes introduced in minor releases, we promise to provide clear migration instructions and possibly tools for automated conversion.
```

Reply:
This paragraph wasn't so much about updating Godot game projects, it's more about using a NuGet package that targets an older version of the _GodotSharp_ assembly.

In this case, if Godot breaks compat, the library will throw exceptions so providing automated conversion or migration instructions are pretty much impossible. The only options that the user has are:

- Stop using the library entirely.
- Wait for the library to release an update that argets a newert version of _GodotSharp_.

So what I wanted to convey with my paragraph was that we'll try to avoid breaking compatibility, so that libraries released _today_ are still usable in a future version of Godot. We obviously can't promise that breaking changes will never happen, but I do hope that we can avoid breaking changes as much as possible (especially after 4.1).

Reply:
Added a note about GodotSharp to hook it back to the original point. (also resolved on chat).